dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java

```diff
@@ -280,22 +279,22 @@ boolean removeTransformer(ClassFileTransformer transformer) {
                     clazz = findLoadedClass(name);
                     if (clazz == null) {
                         ByteResourceInformation byteResInfo = this.findClassBytes(name);
-                        clazz = definePackageAndClass(name, byteResInfo, byteResInfo.getBytes());
+                        if (byteResInfo != null) {
+                            clazz = definePackageAndClass(name, byteResInfo, byteResInfo.getBytes());
+                        } else {
+                            // Check the common libraries.
+                            clazz = findClassCommonLibraryClassLoaders(name);
+                        }
                     }
                 }
-            } catch (ClassNotFoundException cnfe) {
-                // Check the common libraries.
-                clazz = findClassCommonLibraryClassLoaders(name);
             } finally {
                 ThreadIdentityManager.reset(token);
             }
             return clazz;
         }
 
```

Comment:
It is really confusing why synchronization on `getClassLoadingLock(name)` and a call to `findLoadedClass(name)` is not needed here when `transformers.isEmpty()`.  Also missing is `ThreadIdentityManager` calls.  This all seems like it would cause issues.  Some comments must be added to explain this, right now I am convinced something is wrong here.

Reply:
I do understand your concern here.  From looking at the code it appears that we get the lock because of the LinkageError comment in lines 275 to 279.  The main difference between the two is the call to findLoadedClass.  Not sure why a LinkageError is possible without a transformer and it isn't with a transformer.

The findClassBytes method also calls ThreadIdentityManager.  Something seems a little odd here yes.  The paths should be consistent what needs to be guarded with the ThreadIdentityManager.

Can it be deferred to another PR instead of holding this one up?